docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,137 @@
+
+This method requires infrastructure knowledge including image creation, registry access and training job spec. This makes it difficult for data scientists to purely focus at the application layer.
+
+**<h4>Higher level SDK</h4>**
+
+To provide a better user experience, a new higher level SDK was added in[ https://github.com/kubeflow/training-operator/pull/1659](https://github.com/kubeflow/training-operator/pull/1659)  which avoided the need of image creation. Instead of creating the image, we can directly pass the custom script to train_func parameter and final pytorch job is created
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='hf://openchat/openchat_3.5', 
+   dataset= 's3://doc-example-bucket1/train_dataset',
+   eval_dataset = "s3://doc-example-bucket1/eval_dataset"
+   HuggingFace(access_token = "hf_..." , peft_config = {lora_alpha: 16}, transformerClass="Trainer"), 
+   s3(access_token = "s3 access token", region="us-west-2"), 
+   {learning_rate=0.1}
+)
+```
+
+The new proposed API takes following arguments 
+
+1. System parameters - Number of nodes, number of procs per node(GPUs per node).
+2. Model parameters - Model provider and repository details.
+3. Dataset parameters - Dataset provider and dataset details.
+4. Training parameters - Training specific parameters like learning rate etc.
```

Comment:
Do `parameters` mean args in kubernetes? Or any other?

Reply:
I didn't get the question @tenzen-y

Reply:
I meant how we pass the `parameters` to the kubernetes.
Embed it to `{func_code}`? Or pass it to container commands/args?

Reply:
I will pass it in init container args or init container environment variables

Reply:
I see. Thanks for this clarification. Could you mention it in this proposal?

Reply:
done